### DISCUSSION TITLE
Allow compilation with OpenCL?

### DISCUSSION THREAD

**Discussion by ballardtj** [cdriveraus/ctsem#61] [Ideas]
Hi Charles,

I hope all is well. I'm interested in the possibility of being able to run ctsem out of the box on GPU. I believe rstan now can compile with OpenCL. The `stan_model()` has a `use_opencl` argument set to FALSE by default. It looks like additional arguments to `ctStanFit()` can currently be passed to `stan()`, but not to `stan_model()`. How easy would it be to allow the user to add arguments to `ctStanFit()` that can be passed to `stan_model()` so we can compile with OpenCL?

Cheers,
Tim

**Comment by cdriveraus** [comment]
Pretty easy. Line 675 of ctStanFit. I can do it next week.

**Comment by cdriveraus** [comment]
Ok this is done, curious to hear how it works for you!

**Reply by ballardtj** [reply]
Thanks! It looks to have made a huge difference. Reduced my model's run time from between 2.5 to 3.5 hours (depending on the chain) to between 1 and 1.3 hours. Not exactly a rigorous test, but it's definitely promising.

**Reply by cdriveraus** [reply]
nice! Thanks for reporting back. I didn't expect to see such gains since the structure is many small matrix operations that must be performed in sequence, so this is quite interesting. Can you (roughly) give the basic details of your problem, e.g. number of latent states, random effects, number of observed variables, number of data rows?

**Reply by ballardtj** [reply]
Actually, I spoke too soon. It turns out it was more about the particular partition I was using on our HPC than the GPU itself. If I run the model on the same partition, with and without using a GPU, the run time is about the same. If anything, the GPU is slightly slower. Sorry about the false alarm!

**Reply by cdriveraus** [reply]
aha. yes unfortunately this makes more sense ;) I expect some nice speed gains if the code is re-written to use native stan parallelisation rather than the clunky 'split into multiple models and parallelise in R' approach I currently take, but there's a fair bit of complexity to that re-write...
